README.md
H2O - an optimized HTTP server with support for HTTP/1.x and HTTP/2
===

[![Build Status](https://travis-ci.org/h2o/h2o.svg?branch=master)](https://travis-ci.org/h2o/h2o)

Copyright (c) 2014-2016 [DeNA Co., Ltd.](http://dena.com/), [Kazuho Oku](https://github.com/kazuho/), [Tatsuhiko Kubo](https://github.com/cubicdaiya/), [Domingo Alvarez Duarte](https://github.com/mingodad/), [Nick Desaulniers](https://github.com/nickdesaulniers/), [Marc Hörsken](https://github.com/mback2k), [Masahiro Nagano](https://github.com/kazeburo/), Jeff Marrison, [Daisuke Maki](https://github.com/lestrrat/), [Laurentiu Nicola](https://github.com/GrayShade/), [Justin Zhu](https://github.com/zlm2012/), [Tatsuhiro Tsujikawa](https://github.com/tatsuhiro-t), [Ryosuke Matsumoto](https://github.com/matsumoto-r), [Masaki TAGAWA](https://github.com/mochipon), [Masayoshi Takahashi](https://github.com/takahashim), [Chul-Woong Yang](https://github.com/cwyang), [Shota Fukumori](https://github.com/sorah), [Satoh Hiroh](https://github.com/cho45), [Fastly, Inc.](https://www.fastly.com), [David Carlier](https://github.com/devnexen), [Frederik Deweerdt](https://github.com/deweerdt)

H2O is a new generation HTTP server.
Not only is it very fast, it also provides much quicker response to end-users when compared to older generations of HTTP servers.

Written in C and licensed under [the MIT License](http://opensource.org/licenses/MIT), it can also be used as a library.

For more information, please refer to the documentation at [h2o.examp1e.net](https://h2o.examp1e.net).


Changes
Revision history for H2O.

2.1.0 TBD
	- [core] TCP latency optimization #873 #1076 (Kazuho Oku)
	- [core] provide tag to include other YAML files from the configuration file #1022 (Ichito Nagata)
	- [core] accept sequence of mappings for path-level configuration #1042 (Ichito Nagata)
	- [core] fix broken support for TCP Fast Open in OS X #1065 (Ichito Nagata)
	- [access-log] provide directive to emit request-level errors #1075 (Kazuho Oku)
	- [access-log] emit values of all `set-cookie` headers concatenated #1161 (Kazuho Oku)
	- [fastcgi] fix connection failure when `fastcgi.spawn` is used with an uid #1119 (Kazuho Oku)
	- [file] more pre-defined MIME types #1103 (Joe Duarte)
	- [http2][proxy] recognize link rel=preload headers in interim response as a trigger to push resources #916 (Kazuho Oku)
	- [http1][http2] validate characters used in the headers #974 #1044 (Frederik Deweerdt, Kazuho Oku)
	- [http1][http2] notify error downstream when an error occurred while generating a response #1031 (Frederik Deweerdt)
	- [http1][http2] fix resource leak upon upgrade failure to HTTP/2 #1161 (Frederik Deweerdt)
	- [http2] add `http2-push-preload` directive to turn off H2 push being initiated by link rel=preload header #929 (Kazuho Oku)
	- [http2] add support for `cache-digest` header #967 #988 (Kazuho Oku)
	- [http2] drop `host` header in HTTP/2 layer #973 #998 (Frederik Deweerdt, Kazuho Oku)
	- [http2] don't use etag for calculating casper cookie #986 (Kazuho Oku)
	- [http2] add support for H2 debug state #1019 (Ichito Nagata)
	- [mruby] add dos_detector mruby handler #1013 (Ichito Nagata)
	- [mruby] add DSL for access control lists (acl) #1016 (Ichito Nagata)
	- [mruby] share mruby state and constants between handlers #1032 (Ichito Nagata)
	- [mruby] add library for address-block-based access control #1038 (Ichito Nagata)
	- [proxy] add an option to connect to upstream using PROXY protocol #930 (Kazuho Oku)
	- [proxy] don't escape `:` in URI path #977 (Frederik Deweerdt)
	- [proxy] preserve received URLs as much as possible #985 #1071 (Frederik Deweerdt)
	- [proxy] add an option to prevent emiting x-forwarded-* headers #999 (Frederik Deweerdt)
	- [proxy] cache TLS session used for upstream connections #1053 (Ichito Nagata)
	- [proxy] turn on/off on-the-fly compression based on the `x-compress-hint` header #1085 (Frederik Deweerdt)
	- [ssl] set add_lock callback to prevent unnecessary lock-add-unlock #983 (Roberto Guimaraes)
	- [ssl] add support for OpenSSL 1.1.0 #1064 (Kazuho Oku)
	- [status] collect and report HTTP statistics #893 (Frederik Deweerdt)
	- [status] report additional stats when jemalloc is used #1017 (Frederik Deweerdt)
	- [throttle] add new handler for throttling the response bandwidth #917 (Justin Zhu)
	- [libh2o] provide `h2o_rand` that calls the appropriate random function depending on the OS #927 (David CARLIER)
	- [libh2o] do not require use of picohttpparser.h when using the HTTP/1 client #946 (Kazuho Oku)
	- [libh2o] install library files to the correct location #1116 (Frederik Deweerdt)
	- [misc] provide `crash-handler` directive to customize crash logging #935 (Frederik Deweerdt)
	- [misc] guess the default location of h2o.conf #969 (Davsid Carlier)
	- [misc] allow to disable libuv even when it is found #995 (Frederik Deweerdt)
	- [misc] add font/woff2 to the default mime-type mapping #1066 (Andy Davies)
	- [misc] mark JavaScript and JSON files as compressible by default #1067 (Kazuho Oku)

2.0.6 2017-01-05 05:31:00+0000
	- [compress] fix the compression quality being ignored #1154 (Yannick Koechlin)
	- [mruby] stop GIT access during build #1149 (parly)

2.0.5 2016-12-21 06:00:00+0000
	- [security fix] fix use-after-free vulnerability CVE-2016-7835 #1144 (Frederik Deweerdt, Kazuho Oku)
	- [core] fix busy loop after receiving SIGTERM (linux) #1100 (Kazuho Oku, Frederik Deweerdt)
	- [core] don't try to register kevent changes more than once (*BSD, OS X) #1113 (Ichito Nagata)
	- [compress] set `vary: accept-encoding` upon negotiation failure of the compression method #1083 (Frederik Deweerdt)
	- [file] add missing `</ul>` #1106 (Kazuho Oku)
	- [http2] fix a bug that left connections open #1090 (Kazuho Oku)
	- [http2] ignore PRIORITY frames that reference closed pushed streams #1105 (Frederik Deweerdt)
	- [http2] add `Secure` attribute to the casper cookie #1134 (Kazuho Oku)
	- [http2] permit use of HEADERS with a smaller stream ID than a preceding PRIORITY #1136 (Frederik Deweerdt, Kazuho Oku)
	- [mruby] update mruby to HEAD #1135 (Kazuho Oku)
	- [proxy] set `content-length: 0` when receiving a zero-byte POST or PUT #1080 (Frederik Deweerdt)
	- [ssl] update libressl to 2.4.4 #1127 (Kazuho Oku)
	- [ssl] erase OCSP stapling data when the stapling updater returns a permanent failure #1117 (Kazuho Oku)

2.0.4 2016-09-14 08:00:00+0000
	- [security fix][core] fix DoS attack vector CVE-2016-4864 #1077 (Frederik Deweerdt, Kazuho Oku)
	- [libh2o] fix crash on connect timeout #960 (disigma)

2.0.3 2016-09-07 22:03:00+0000
	- [file] don't use `readdir_r` on Linux, Solaris #1046 #1052 (Frederik Deweerdt, Kazuho Oku)
	- [http2] fix negative error code sent when cancelling a pushed stream #1039 (Frederik Deweerdt)
	- [http2] fix a bug that may cause a stream to stall #1040 (Frederik Deweerdt)
	- [http2] fix a bug that reset the stream when receiving HEADERS after PRIORITY #1043 (Frederik Deweerdt)
	- [mruby] fix mruby handler becoming unusable after failed connection in http_request on FreeBSD #1062 (Kazuho Oku)

2.0.2 2016-08-01 20:36:00+0000
	- [fastcgi] setenv should displace HTTP headers #996 (Kazuho Oku)
	- [http2] fix buffer overrun #972 (Frederik Deweerdt)
	- [misc] fix build error when libuv is not found #1008 (nextgenthemes)
	- [misc] fix assertion failure when YAML alias and merge is used in certain way #1011 (Kazuho Oku)

2.0.1 2016-06-23 22:00:00+0000

misc/tokens.pl
#! /usr/bin/perl

# Copyright (c) 2014 DeNA Co., Ltd.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.

use strict;
use warnings;
use List::Util qw(max);
use List::MoreUtils qw(uniq);
use Text::MicroTemplate qw(render_mt);

use constant LICENSE => << 'EOT';
/*
 * Copyright (c) 2014 DeNA Co., Ltd.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */
EOT

my %tokens;
my @hpack;

while (my $line = <DATA>) {
    chomp $line;
    my ($hpack_index, $proxy_should_drop, $is_init_header_special, $http2_should_reject, $copy_for_push_request, $name, $value) =
        split /\s+/, $line, 7;
    next unless $name ne '';
    $tokens{$name} = [ $hpack_index, $proxy_should_drop, $is_init_header_special, $http2_should_reject, $copy_for_push_request ]
        unless defined $tokens{$name};
    if ($hpack_index != 0) {
        $hpack[$hpack_index - 1] = [ $name, $value ];
    }
}

my @tokens = map { [ $_, @{$tokens{$_}} ] } uniq sort keys %tokens;

# generate token.h
open my $fh, '>', 'include/h2o/token.h'
    or die "failed to open include/h2o/token.h:$!";
print $fh render_mt(<< 'EOT', \@tokens, LICENSE)->as_string;
? my ($tokens, $license) = @_;
<?= $license ?>
/* DO NOT EDIT! generated by tokens.pl */
#ifndef h2o__token_h
#define h2o__token_h

? for my $i (0..$#$tokens) {
#define <?= normalize_name($tokens->[$i][0]) ?> (h2o__tokens + <?= $i ?>)
? }

#endif
EOT
close $fh;

# generate token_table.h
open $fh, '>', 'lib/core/token_table.h'
    or die "failed to open lib/core/token_table.h:$!";
print $fh render_mt(<< 'EOT', \@tokens, LICENSE)->as_string;
? my ($tokens, $license) = @_;
<?= $license ?>
/* DO NOT EDIT! generated by tokens.pl */
h2o_token_t h2o__tokens[] = {
? for my $i (0..$#$tokens) {
    { { H2O_STRLIT("<?= $tokens->[$i][0] ?>") }, <?= join(", ", map { $tokens->[$i][$_] } (1..$#{$tokens->[$i]})) ?> }<?= $i == $#$tokens ? '' : ',' ?>
? }
};
size_t h2o__num_tokens = <?= scalar @$tokens ?>;

const h2o_token_t *h2o_lookup_token(const char *name, size_t len)
{
    switch (len) {
? for my $len (uniq sort { $a <=> $b } map { length $_->[0] } @$tokens) {
    case <?= $len ?>:
        switch (name[<?= $len - 1 ?>]) {
?  my @tokens_of_len = grep { length($_->[0]) == $len } @$tokens;
?  for my $end (uniq sort map { substr($_->[0], length($_->[0]) - 1) } @tokens_of_len) {
        case '<?= $end ?>':
?   my @tokens_of_end = grep { substr($_->[0], length($_->[0]) - 1) eq $end } @tokens_of_len;
?   for my $token (@tokens_of_end) {
            if (memcmp(name, "<?= substr($token->[0], 0, length($token->[0]) - 1) ?>", <?= length($token->[0]) - 1 ?>) == 0)
                return <?= normalize_name($token->[0]) ?>;
?   }
            break;
?  }
        }
        break;
? }
    }

    return NULL;
}
EOT
close $fh;

# generate hpack_static_table.h
open $fh, '>', 'lib/http2/hpack_static_table.h'
    or die "failed to open lib/hpack_static_table.h:$!";
print $fh render_mt(<< 'EOT', \@hpack, LICENSE)->as_string;
? my ($entries, $license) = @_;
<?= $license ?>
/* automatically generated by tokens.pl */

static const struct st_h2o_hpack_static_table_entry_t h2o_hpack_static_table[<?= scalar @$entries ?>] = {
? for my $i (0..$#$entries) {
    { <?= normalize_name($entries->[$i][0]) ?>, { H2O_STRLIT("<?= $entries->[$i][1] || "" ?>") } }<?= $i == $#$entries ? "" : "," ?>
? }
};
EOT
close $fh;

sub normalize_name {
    my $n = shift;
    $n =~ s/^://;
    $n =~ s/-/_/g;
    $n =~ tr/a-z/A-Z/;
    "H2O_TOKEN_$n";
}

__DATA__
1 0 0 0 0 :authority
2 0 0 0 0 :method GET
3 0 0 0 0 :method POST
4 0 0 0 0 :path /
5 0 0 0 0 :path /index.html
6 0 0 0 0 :scheme http
7 0 0 0 0 :scheme https
8 0 0 0 0 :status 200
9 0 0 0 0 :status 204
10 0 0 0 0 :status 206
11 0 0 0 0 :status 304
12 0 0 0 0 :status 400
13 0 0 0 0 :status 404
14 0 0 0 0 :status 500
15 0 0 0 1 accept-charset
16 0 0 0 1 accept-encoding gzip, deflate
17 0 0 0 1 accept-language
18 0 0 0 0 accept-ranges
19 0 0 0 1 accept
20 0 0 0 0 access-control-allow-origin
21 0 0 0 0 age
22 0 0 0 0 allow
23 0 0 0 0 authorization
24 0 0 0 0 cache-control
25 0 0 0 0 content-disposition
26 0 0 0 0 content-encoding
27 0 0 0 0 content-language
28 0 1 0 0 content-length
29 0 0 0 0 content-location
30 0 0 0 0 content-range
31 0 0 0 0 content-type
32 0 0 0 0 cookie
33 1 0 0 0 date
34 0 0 0 0 etag
35 0 1 0 0 expect
36 0 0 0 0 expires
37 0 0 0 0 from
38 0 1 1 0 host
39 0 0 0 0 if-match
40 0 0 0 0 if-modified-since
41 0 0 0 0 if-none-match
42 0 0 0 0 if-range
43 0 0 0 0 if-unmodified-since
44 0 0 0 0 last-modified
45 0 0 0 0 link
46 0 0 0 0 location
47 0 0 0 0 max-forwards
48 1 0 0 0 proxy-authenticate
49 1 0 0 0 proxy-authorization
50 0 0 0 0 range
51 0 0 0 0 referer
52 0 0 0 0 refresh
53 0 0 0 0 retry-after
54 1 0 0 0 server
55 0 0 0 0 set-cookie
56 0 0 0 0 strict-transport-security
57 1 1 1 0 transfer-encoding
58 0 0 0 1 user-agent
59 0 0 0 0 vary
60 0 0 0 0 via
61 0 0 0 0 www-authenticate
0 1 0 1 0 connection
0 0 0 0 0 x-reproxy-url
0 1 1 1 0 upgrade
0 1 0 1 0 http2-settings
0 1 0 1 0 te
0 1 0 0 0 keep-alive
0 0 0 0 0 x-forwarded-for
0 0 0 0 0 x-traffic
0 0 0 0 0 cache-digest
0 0 0 0 0 x-compress-hint
